include/libkrun.h

```diff
@@ -113,6 +122,49 @@ int32_t krun_set_data_disk(uint32_t ctx_id, const char *disk_path);
  */
 int32_t krun_add_disk(uint32_t ctx_id, const char *block_id, const char *disk_path, bool read_only);
 
+/* Supported disk image formats */
+#define KRUN_DISK_FORMAT_RAW 0
+#define KRUN_DISK_FORMAT_QCOW2 1
+/**
+ * Adds a disk image to be used as a general partition for the microVM. The supported
+ * image formats are: "raw" and "qcow2".
+ *
+ * This API is mutually exclusive with the deprecated krun_set_root_disk and
+ * krun_set_data_disk methods and must not be used together.
+ *
+ * SECURITY NOTE:
+ * Libkrun is not responsible for probing the disk image provided, as probing a disk image should
+ * be avoided. However, if one must probe the disk image, it falls onto the user. The disk image
+ * must only be probed once, and the probe of the image must be done before the first boot of the
+ * microVM. 
```

Comment:
I find it too easy to read the first part and get the idea that probing “‘is to be avoided’, but, well, avoiding doesn’t mean ‘must not’, right?”

Also, the concrete problem is buried below, and it reads a bit like qcow2 images referencing other files is only a problem if the guest modified a raw image to do so; but actually it’s a general security concern, for any image that isn’t fully trusted, so probably should be the very first thing mentioned.

(I didn’t really want to write a full proposal, but it kind of happened accidentally, so… What would you think about the following?)

___

“Non-Raw images can reference other files, which libkrun will automatically open, and to which the guest will have access. Libkrun should therefore never be asked to open an image in a non-Raw format when it does not come from a fully trustworthy source.

Consequently, probing an image’s format is quite dangerous and to be avoided if at all possible, which is why libkrun provides no facilities for doing so. If it is not clear what format an image has, it may also not be clear whether it can be trusted not to reference files to which the guest should not have access.

If probing absolutely cannot be avoided, it must only be done on images that are fully trusted, i.e. before a potentially untrusted guest had write access to it. Specifically, consider that a guest has full access to all of a Raw image, and can therefore turn it into a file in an arbitrary format, for example, into a Qcow2 image, referencing and granting a malicious guest access to arbitrary files. To hand a Raw image to an untrusted and potentially malicious guest, and then to re-probe it after the guest was able to write to it (when it can no longer be trusted), would therefore be a severe security vulnerability.

Therefore, after having probed a yet fully trusted image once, the result must be remembered so the image will from then on always be opened in the format that was detected originally. When adhering to this, a guest can write anything they want into a Raw image, it is always going to be opened as a Raw image, preventing the security vulnerability outlined above.

However, if at all possible, the image format should be explicitly selected based on knowledge obtained separately from the pure image data, for example by the user.”

Reply:
My suggestion may well be (much?) too long, but I think it is important to be specific about the exact problem (formatted images give guests access to other files) so that users always consider this in all of its implications (i.e. not only get format detection right, but also ensure that even a freshly downloaded image won’t contain malicious file references in the first place).

Reply:
Just added your suggestions